File: Kth Largest Sum in a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long kthLargestLevelSum(TreeNode* root, int k) {
        vector<int64_t> summ;
        function<void(TreeNode*, int)> recur;
        recur = [&recur, &summ] (TreeNode* node, int level) {
            if (!node) return;

            if (summ.size() == level) summ.push_back(node->val);
            else summ[level] += node->val;
            recur(node->left, level+1);
            recur(node->right, level+1);
        };
        recur(root, 0);
        if (summ.size() < k) return -1;
        sort(summ.begin(), summ.end());
        return summ[summ.size()-k];
    }
};

File: Kth Largest Sum in a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthLargestLevelSum(self, root: Optional[TreeNode], k: int) -> int:
        summ = []

        def recur(node, level):
            nonlocal summ
            if not node:
                return

            if len(summ) == level:
                summ.append(node.val)
            else:
                summ[level] += node.val
            recur(node.left, level + 1)
            recur(node.right, level + 1)

        recur(root, 0)
        if len(summ) < k:
            return -1
        summ.sort()
        return summ[-k]
